languages/rust/generic/Parse_rust_tree_sitter.ml

```diff
@@ -82,11 +82,46 @@ type rust_macro_item =
   | MacTree of rust_macro_item list G.bracket
   | MacTreeBis of rust_macro_item list G.bracket * G.ident option * G.tok
 
-(* Note that the commas are considered like any other tokens in a Rust macro;
- * they are not separators between rust_macro_items.
- *)
 let rec macro_items_to_anys (xs : rust_macro_item list) : G.any list =
-  xs |> Common.map macro_item_to_any
+  (* Note that the commas are considered like any other tokens in a Rust macro;
+     * they are not separators between rust_macro_items.
+  *)
+  (* Thus, we need this machinery in order to parse a list of macro items
+     to see if the macro invocation looks like a normal expression. For instance,
+     if a sequence of macro items looks like
+     <expr> , <expr> , <expr>
+     then we assume that this is something which looks like a series of
+     arguments, so we produce a single `Any`, which is `Args`.
+     Note that <expr> can also occur once with no commas, or zero times.
+  *)
+  let macro_item_to_arg = function
+    | MacAny (G.E e) -> Some (G.Arg e)
+    | MacAny (G.I e) -> Some (G.Arg (G.N (G.Id (e, G.empty_id_info ())) |> G.e))
+    | MacAny (G.Ar arg) -> Some arg
+    | MacAny _
+    | MacTreeBis _
+    | MacTree _ ->
+        None
+  in
+  let rec try_as_normal_args = function
+    | [] -> Some []
+    | [ mac ] ->
+        let* arg = macro_item_to_arg mac in
+        Some [ arg ]
+    (* For now, we just directly case on the token to see if its a comma.
+       This is a fragile approach, but we're close to the origin of the
```

Comment:
In which sense it is "fragile", and what does "relatively fine" mean?

Reply:
I suppose I mean "fragile", as in there could be weird cases where the string of the token is not exactly `","`, because I don't fully trust pattern matching on the string inside the token. I don't know that for sure, but this is just less statically safe to my mind. "Relatively fine", as in, there's few opportunities for the string inside the token to change, and there's not much else we could have done prior to this to make this logic easier.

Reply:
I can leave a comment